www/js/app.js

```diff
@@ -1253,9 +1252,25 @@ define(['jquery', 'zimArchiveLoader', 'uiUtil', 'settingsStore','abstractFilesys
             if (regexpActiveContent.test(htmlArticle)) uiUtil.displayActiveContentWarning();
         }
 
+        // Calculate the current article's ZIM baseUrl to use when processing relative links
+        var baseUrl = dirEntry.namespace + '/' + dirEntry.url.replace(/[^/]+$/, '');
+
         // Replaces ZIM-style URLs of img, script, link and media tags with a data-kiwixurl to prevent 404 errors [kiwix-js #272 #376]
-        // This replacement also processes the URL to remove the path so that the URL is ready for subsequent jQuery functions
-        htmlArticle = htmlArticle.replace(regexpTagsWithZimUrl, '$1data-kiwixurl$2');
+        // This replacement also processes the URL relative to the page's ZIM URL so that we can find the ZIM URL of the asset
+        // with the correct namespace (this works for old-style -,I,J namespaces and for new-style C namespace)
+        htmlArticle = htmlArticle.replace(regexpTagsWithZimUrl, function(m0, m1, m2, relAssetUrl) {
+            var assetZIMUrl = uiUtil.deriveZimUrlFromRelativeUrl(relAssetUrl, baseUrl);
+            // Uncomment logging below to test the calculation of relative URLs if you are having issues 
+            /**
+                console.log('ZIM URL: ' + dirEntry.namespace + '/' + dirEntry.url);
+                console.log('Full URL: ' + relAssetUrl);
+                console.log('Base URL: ' + baseUrl);
+                console.log('Calculated asset URL: ' + decAssetZIMUrl);
+            **/
+            // DEV: Note that deriveZimUrlFromRelativeUrl produces a *decoded* URL (and removes any URI component)
+            // so we re-encode it with encodeURI (this does not encode forward slashes) 
+            return m1 + 'data-kiwixurl' + m2 + encodeURI(assetZIMUrl);
```

Comment:
Query: As noted in the comment, `deriveZimUrlFromRelativeUrl` has the side effect of discarding any URI component. Is this an issue for assets? We are only here processing links for tags `img|script|link|track` in jQuery mode. We don't support script in any case (where query string could be important). I think it's not relevant for css. It would only be relevant for images if some processing were being done on the image, which we don't do in jQuery mode. We could save the query string in this function and add it back here, but is it necessary?

Reply:
Actually, I've now resolved this simply by not capturing the query string or hash part of the URL before processing and adding back into the document. This is a better solution.